### ISSUE TITLE
He et al "Delving deep into rectifiers" initialization

### ISSUE THREAD

**Issue by dwf** [mila-udem/blocks#1051]
We should really have this.

One issue is that you need to be able to a) inspect the Brick that the parameter belongs to (not a problem with annotations) and b) calculate a given set of weights' "fan-in". 

I'd rather make this an annotation you can just look up on the variable rather than hard-code something in the Initialization object that is bound to break.

**Comment by sisp** [comment]
Can you elaborate on a) and b)? Why do you need to know the brick? To compute the "fan-in", you get the ``shape`` argument in `:class:~blocks.initialize.NdarrayInitialization.generate` which contains the "fan-in" for fully connected layers in ``shape[0]`` and for convolutional layers in ``np.prod(shape[1:])``.

**Comment by rizar** [comment]
Yep, I also do not understand why fan-in is not a part of the parameter's
shape.

**Comment by dwf** [comment]
It's not very future-proof to take a shape-tuple and do some prespecified
operation on it without knowing what the semantics that the source Brick
assigns to it are. What I'm saying is that the brick should make this
explicit somehow, either by exposing a method or by annotating the variable.

**Comment by rizar** [comment]
Actually you are right, since we even don't have an expression that works for fully-connected and convolutional layers, let alone all possible crazy layers of the future. 

We can have the following-interfaces: 

```
class Brick(object):

def parameter_properties(self): 
    """Return a dictionary of parameter properties."""
    pass

class Linear(...):

def parameter_properties(self):
    return {'W' : {'fan_in': self.input_dim}}

```

These properties can be automatically passed further in the new implementation of `_initialize` that @memimo  is working on.
